Support for N-ary operations and common intrinsics (towards 339)

arporter opened the pull request:
This is my second attempt (was previously #417).

arporter commented:
I think is this is now ready for first review from either @sergisiso or @rupertford.

sergisiso commented:
@arporter  Overall the PR looks good, I just made a few in-line comments. Tests pass but travis hasn't generated a coverage report, if there isn't one after addressing the current comments I will check manually.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -119,6 +115,7 @@ def colored(text, _):
                        "Reference": "yellow",
                        "BinaryOperation": "blue",
                        "UnaryOperation": "blue",
+                       "NaryOperation": "blue",
```
Oh yes, that makes more sense.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6080,38 +6095,23 @@ def function_format(function_str, expr_str):
     Operator = Enum('Operator', [
         # Arithmetic Operators
-        'ADD', 'SUB', 'MUL', 'DIV', 'REM', 'POW',
+        'ADD', 'SUB', 'MUL', 'DIV', 'REM', 'POW', 'MODULUS', 'SUM',
```
I didn't even know about MODULO. Changed back to REM!

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6410,7 +6407,21 @@ class Fparser2ASTProcessor(object):
         ('-', UnaryOperation.Operator.MINUS),
-        ('.not.', UnaryOperation.Operator.NOT)])
+        ('.not.', UnaryOperation.Operator.NOT),
+        ("abs", UnaryOperation.Operator.ABS),
```
Fair point - changed.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6431,7 +6442,17 @@ class Fparser2ASTProcessor(object):
+        ('sum', BinaryOperation.Operator.SUM),
+        ('max', BinaryOperation.Operator.MAX),
+        ('min', BinaryOperation.Operator.MIN),
+        ('mod', BinaryOperation.Operator.MODULUS)])
```
Done.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -7334,21 +7356,37 @@ def _unary_op_handler(self, node, parent):
         Transforms an fparser2 UnaryOpBase to the PSyIR representation.
```
Oops, well spotted.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -7357,31 +7395,145 @@ def _unary_op_handler(self, node, parent):
+        '''
+        Transforms an fparser2 Intrinsic_Function_Reference to the PSyIR
+        representation. Since Fortran Intrinsics can be unary, binary or
+        nary this handler identifies the appropriate 'sub handler'
```
Oh yes. Fixed.

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -128,6 +128,27 @@ def gen_kind(symbol):
+    :param op_map: mapping from string representation of operator to \
+                   enumerated type.
+    :type op_map: collections.OrderedDict
+
```
Done.

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -264,31 +285,46 @@ def binaryoperation_node(self, node):
             # This is a binary operation
+            if oper.upper() in FORTRAN_INTRINSICS:
+                # This is a binary intrinsic function.
+                return "{0}({1},{2})".format(oper.upper(), lhs, rhs)
```
Yes. Done.

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -264,31 +285,46 @@ def binaryoperation_node(self, node):
+            if oper.upper() in FORTRAN_INTRINSICS:
+                # This is a binary intrinsic function.
+                return "{0}({1},{2})".format(oper.upper(), lhs, rhs)
             return "{0}{1}{2}".format(lhs, oper, rhs)
```
Done.

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -264,31 +285,46 @@ def binaryoperation_node(self, node):
+            arg_list.append(self._visit(child))
+        try:
+            oper = mapping[node.operator]
+            return "{0}({1})".format(oper.upper(), ",".join(arg_list))
```
Yes, done.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -4614,20 +4644,42 @@ def test_fparser2astprocessor_handling_part_ref(f2008_parser):
+        ('x = int(a, 8)', CodeBlock, None),
+        ('x = log(a)', UnaryOperation, UnaryOperation.Operator.LOG),
+        ('x = log10(a)', UnaryOperation, UnaryOperation.Operator.LOG10),
+        ('x = mod(a, b)', BinaryOperation, BinaryOperation.Operator.MODULUS),
```
Done.

arporter commented:
All points addressed. Ready for next review.
(I think CodeCov is not reporting because it doesn't have the coverage of the point on master from which this branch was taken. I think that's because I was over-zealous in telling GitHub not to trigger Travis when doing the update of the changelog/top-level user guide in the final step of merging a branch to master. Lesson learned!)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -46,6 +46,7 @@
 import six
 from psyclone.configuration import Config
 from psyclone.core.access_type import AccessType
+from fparser.two import Fortran2003
```
Mine didn't but fixed :-)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -5938,67 +5933,48 @@ def gen_c_code(self, indent=0):
-        super(UnaryOperation, self).__init__(parent=parent)
+        super(Operation, self).__init__(parent=parent)
 
         if not isinstance(operator, self.Operator):
```
Good idea, done.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -5938,67 +5933,48 @@ def gen_c_code(self, indent=0):
-
         :param int indent: level to which to indent output.
         '''
         print(self.indent(indent) + self.coloured_text + "[operator:'" +
```
Good point, done.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6080,38 +6093,23 @@ def function_format(function_str, expr_str):
-    :param operator: Enumerated type capturing the binary operator.
-    :type operator: :py:class:`psyclone.psyGen.BinaryOperation.Operator`
-    :param parent: the parent node of this BinaryOperation in the PSyIR.
-    :type parent: :py:class:`psyclone.psyGen.Node`
```
OK, done.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6225,6 +6195,30 @@ def function_format(function_str, expr1, expr2):
+    Node representing a n-ary operation expression. The n operands are the
+    stored as the 0 - n-1th children of this node and the type of the operator
+    is held in an attribute.
+
```
Done.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6225,6 +6195,30 @@ def function_format(function_str, expr1, expr2):
+
 class Array(Reference):
     '''
     Node representing an Array reference. As such it has a reference and a
```
Done.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -4637,18 +4689,130 @@ def test_fparser2astprocessor_handling_intrinsics(f2008_parser):
+    parse tree has an unexpected structure. This is a hard error to
+    provoke since fparser checks that the number of arguments is correct. '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Execution_Part, Name
```
Bizarrely, `Execution_Part` *is* used but `Name` isn't so I removed that one :-)

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -4637,18 +4689,130 @@ def test_fparser2astprocessor_handling_intrinsics(f2008_parser):
+    ''' Check that we correctly handle nested intrinsic functions. '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Execution_Part
+    from psyclone.psyGen import Reference
```
Removed.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -4637,18 +4689,130 @@ def test_fparser2astprocessor_handling_intrinsics(f2008_parser):
+    ''' Check that we correctly handle array products. '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Execution_Part
+    from psyclone.psyGen import CodeBlock
```
Removed.

arporter commented on src/psyclone/tests/psyir/backend/fortran_test.py:
```diff
@@ -317,14 +317,15 @@ def test_fw_binaryoperator():
     exception if an unknown binary operator is found.
 
     '''
+    from psyclone.psyGen import Fparser2ASTProcessor
```
Oops, apologies.

arporter commented:
I've added some documentation and fixed all the other issues (apologies for missing the pylint stuff). As discussed, the auto-generated doc for the enumerations isn't great but I propose to live with it for the time being as we don't want to manually replicate the lists in the documentation.
